Add SNP platform status and guest launch ioctls

tylerfanelli opened the pull request:
Imported from the `snp` crate.

haraldh committed:
feat: add KVM_MEMORY_ENCRYPT_REG_REGION and UNREG

Both are specified wrongly as `_IOR` instead of `_IOW`.

(backported from enarx/sev-iocuddle commit 4ba150be7dd5f69017670e5f1697828aadea1e0a)
[ Connor Kuehl: destination file is different in this repo, so adjusted
  patch. Also squashed in the follow-up patch that fixes the
  capitalization of the type. ]

Signed-off-by: Connor Kuehl <ckuehl@redhat.com>

tylerfanelli committed:
firmware: Add SNP platform status ioctl

Signed-off-by: Tyler Fanelli <tfanelli@redhat.com>

tylerfanelli committed:
kvm: launch: Add SEV/KVM guest launch ioctls from snp crate

'kvm_ioctls' and 'kvm_bindings' were both bumped to their latest versions to avoid a naming conflict
in the kvm_userspace_memory_region struct of earlier versions.

Signed-off-by: Tyler Fanelli <tfanelli@redhat.com>

haraldh reviewed (changes requested):
For the tests to succeed, you should also change:
```diff
-    runs-on: [self-hosted, linux, sev]
+    runs-on: [self-hosted, linux, snp]
```
in https://github.com/enarx/sev/blob/3a7775347fec52c335dbeb15cd3da3fc965e6949/.github/workflows/test.yml#L34

connorkuehl commented on src/firmware/linux/mod.rs:
```diff
@@ -118,6 +118,56 @@ impl Firmware {
+        &mut self,
+        config_address: u64,
+        certs_address: u64,
+        certs_len: u32,
```
This is not a safe interface. It can be made safe by taking a `&Config` (or whatever the input type best expresses what the firmware wants) and a `&Chain` (or whatever input certificate/chain the firmware wants). Then, internally, I'd suggest updating `SnpSetExtConfig::new` to take these references. `SnpSetExtConfig` can use `std::marker::PhantomData` to describe the lifetimes of these inputs and internally may case the references to their u64 representations of their addresses.

This way, the caller does not have to dip their toes into unsafety, and at the same time, the lifetime of these memory addresses can be enforced automatically.

connorkuehl commented on src/firmware/mod.rs:
```diff
@@ -240,6 +240,17 @@ pub enum State {
+        let state = match self {
+            State::Uninitialized => "Uninitialized",
+            State::Initialized => "Initialized",
+            State::Working => "Working",
```
Lowercase please. Callers can title case or whatever-case the strings if they want to.

connorkuehl commented on src/firmware/types.rs:
```diff
@@ -165,3 +165,72 @@ impl<'a> GetId<'a> {
+            config_address,
+            certs_address,
+            certs_len,
+            _phantom: PhantomData,
```
This phantom data doesn't appear to be doing anything for us. See my comment above regarding safe interfaces for this.

tylerfanelli replied:
Yea, I see what you mean. I can pass `&Config` and `&Chain` in order to make the lifetimes match `SnpSetExtConfig` via `PhantomData`.

connorkuehl commented on src/lib.rs:
```diff
@@ -198,3 +198,36 @@ impl TryFrom<&sev::Chain> for Generation {
+    fn encode(&self, mut writer: impl Write, _: ()) -> std::io::Result<()> {
+        writer.save(self)
+    }
+}
```
Why are you implementing codicon traits for this?

tylerfanelli replied:
I originally seen codicon traits implemented for the original `Build` and wanted `SnpBuild` to mirror the `Build` functionality. I'll remove them.

connorkuehl replied:
iirc those were implemented to serialize/deserialize data to persistent storage (file) or over the wire since they're used in remote attestation messages.

connorkuehl commented on tests/api.rs:
```diff
@@ -119,3 +119,42 @@ fn get_identifier() {
+        certs_address: {}\n
+        certs_len: {}",
+        config.config_address, config.certs_address, config.certs_len
+    );
```
Does this affect the entire system? I wonder if this should instead be gated behind a dangerous_hw_tests feature.

tylerfanelli replied:
Can you further explain?

connorkuehl replied:
For example, the reset command is gated behind a hardware feature called `dangerous_hw_tests` which must be manually enabled because it clears the persistent state of the platform. If this happens while CI is running, some of the tests will spuriously fail.

Do you anticipate running this command to have similar side effects against other VMs?

tylerfanelli replied:
It seems that randomly setting new configuration values via `snp_set_ext_config` could affect other VMs (at the very least, ones that are running SNP) on a system. I'll add a dangerous_hw_tests feature.

connorkuehl commented on src/kvm/types.rs:
```diff
@@ -111,3 +112,134 @@ impl<'a> LaunchMeasure<'a> {
+        Self {
+            uaddr: data.as_ptr() as _,
+            len: data.len() as _,
+            imi_page: update.imi_page as _,
```
```suggestion
            imi_page: if update.imi_page { 1 } else { 0 },
```

connorkuehl commented on src/launch/mod.rs:
```diff
@@ -192,3 +194,171 @@ impl codicon::Encoder<()> for Measurement {
+        const MIGRATE_MA =  1 << 2;
+
+        /// If enabled, debugging is allowed.
+        const DEBUG =       1 << 3;
```
Weird spacing.

connorkuehl commented on src/launch/mod.rs:
```diff
@@ -192,3 +194,171 @@ impl codicon::Encoder<()> for Measurement {
+
+    /// Hypervisor provided value to indicate guest OS visible workarounds.The format is hypervisor defined.
+    pub gosvw: [u8; 16],
+}
```
This is unsafe. Please use PhantomData to bind this with the lifetime that describes the memory address.

connorkuehl commented on src/launch/mod.rs:
```diff
@@ -192,3 +194,171 @@ impl codicon::Encoder<()> for Measurement {
+
+/// Encapsulates the various data needed to begin the update process.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SnpUpdate {
```
This type could use some constructor functions, otherwise a caller is going to have to piece this together every time on their own.

connorkuehl commented on src/launch/mod.rs:
```diff
@@ -192,3 +194,171 @@ impl codicon::Encoder<()> for Measurement {
+        const EXECUTE_USER =        1 << 2;
+
+        /// Page is executable by the VMPL in CPL2, CPL1, and CPL0.
+        const EXECUTE_SUPERVISOR =  1 << 3;
```
Weird spacing. Generally when one puts identifiers and values in columns like this, the `=` is one space away from the value.

connorkuehl commented on src/launch/mod.rs:
```diff
@@ -192,3 +194,171 @@ impl codicon::Encoder<()> for Measurement {
+    pub id_block_uaddr: u64,
+
+    /// Userspace address of the authentication information of the ID block. Ignored if ID_BLOCK_EN is 0.
+    pub id_auth_uaddr: u64,
```
These addresses are unsafe. Please use a constructor function that takes a reference to the type and put a phantomdata in the struct so the struct can ensure its lifetime is valid for the lifetime of these addresses.

connorkuehl commented on Cargo.toml:
```diff
@@ -37,7 +37,7 @@ iocuddle = "0.1"
 [dev-dependencies]
 dirs = "3.0"
-kvm-bindings = "0.4"
+kvm-bindings = "0.5"
```
I think this should be merged into whichever commit depends on this change, as well as a note describing why you bumped the version.

tylerfanelli replied:
I also bumped `kvm-ioctls` in this commit too, as to avoid the naming conflict between `kvm-bindings` and `kvm-ioctls`

connorkuehl commented on .github/workflows/test.yml:
```diff
@@ -31,7 +31,7 @@ jobs:
   hw:
     name: hw ${{ matrix.toolchain }} ${{ matrix.profile.name }} ${{ matrix.features }}
-    runs-on: [self-hosted, linux, sev]
+    runs-on: [self-hosted, linux, snp]
```
I think this should be merged into the first commit that started testing the SNP ioctls.

haraldh replied:
Somehow this changed to `sev-snp`

```suggestion
    runs-on: [self-hosted, linux, sev-snp]
```

connorkuehl commented on src/kvm/types.rs:
```diff
@@ -111,3 +112,134 @@ impl<'a> LaunchMeasure<'a> {
+            gosvw: start.gosvw,
+            _phantom: PhantomData,
+        }
+    }
```
Why do some parameters live in `SnpStart` but something like the migration agent slice is passed in as a separate argument to this function?

connorkuehl commented on src/kvm/types.rs:
```diff
@@ -111,3 +112,134 @@ impl<'a> LaunchMeasure<'a> {
+            vmpl2_perms: update.vmpl2_perms.bits(),
+            vmpl1_perms: update.vmpl1_perms.bits(),
+            _phantom: PhantomData,
+        }
```
Same question about the split data and SnpUpdate parameters

connorkuehl commented on src/kvm/types.rs:
```diff
@@ -111,3 +112,134 @@ impl<'a> LaunchMeasure<'a> {
+
+/// Complete the guest launch flow.
+#[repr(C)]
+pub struct SnpLaunchFinish<'a> {
+    /// Userspace address of the ID block. Ignored if ID_BLOCK_EN is 0.
+    id_block_uaddr: u64,
+
+    /// Userspace address of the authentication information of the ID block. Ignored if ID_BLOCK_EN is 0.
+    id_auth_uaddr: u64,
+
+    /// Indicates that the ID block is present.
+    id_block_en: u8,
+
+    /// Indicates that the author key is present in the ID authentication information structure.
+    /// Ignored if ID_BLOCK_EN is 0.
+    auth_key_en: u8,
+
+    /// Opaque host-supplied data to describe the guest. The firmware does not interpret this value.
+    host_data: [u8; KVM_SEV_SNP_FINISH_DATA_SIZE],
+
+    _phantom: PhantomData<&'a [u8]>,
+}
```
Should there be two lifetimes? One for the id_block addr and one for the id_auth addr?

connorkuehl replied:
@tfanelli-rh have you looked into this?

tylerfanelli replied:
`SnpFinish`'s lifetime depends on both `id_block` and `id_auth`, and `SnpLaunchFinish`'s lifetime is dependent on `SnpFinish`'s lifetime.

```
pub struct SnpFinish<'a, 'b> {

    pub id_block: Option<&'a [u8]>,

    pub id_auth: Option<&'b [u8]>,

    ......
```

So wouldn't setting `SnpLaunchFinish`'s lifetime to match `SnpFinish`'s preserve the references and make this safe?

connorkuehl replied:
Ohh, I see. Yes, I think you're right.

connorkuehl commented on src/kvm/types.rs:
```diff
@@ -111,3 +112,134 @@ impl<'a> LaunchMeasure<'a> {
+            host_data: finish.host_data,
+            _phantom: PhantomData,
+        }
+    }
```
Same question about split slices and parameters here.

connorkuehl commented on src/launch/mod.rs:
```diff
@@ -192,3 +194,195 @@ impl codicon::Encoder<()> for Measurement {
+
+    /// Hypervisor provided value to indicate guest OS visible workarounds.The format is hypervisor defined.
+    pub gosvw: [u8; 16],
+}
```
I think this should have a constructor function. `ma_en` can be deduced by passing `Option<&[u8]>` for the region of the address space that contains the migration agent.

If it's `Some(blah)` then ma_en = true, otherwise, false.

connorkuehl commented on src/launch/mod.rs:
```diff
@@ -192,3 +194,195 @@ impl codicon::Encoder<()> for Measurement {
+    /// Opaque host-supplied data to describe the guest. The firmware does not interpret this
+    /// value.
+    pub host_data: [u8; KVM_SEV_SNP_FINISH_DATA_SIZE],
+}
```
Similar comment, the constructor function should take Option<&[u8]> and the bools can be deduced by whether or they are Some(&[u8]) for each of the related address slices.

Commit:
Pin encrypted pages by default

Also upgrade kvm-bindings, since it seems semver has been broken for one
of those crates. Otherwise, the build will fail.

While we're in there, also #[allow(dead_code)] for the unpin ioctl. A
future patch may remove the need for that directive once a higher-level
Rust API wraps the pin/unpin functionality, but until then, no need for
the warning.

Signed-off-by: Connor Kuehl <ckuehl@redhat.com>

connorkuehl commented on src/launch/launcher.rs:
```diff
@@ -93,6 +134,31 @@ impl<'a, U: AsRawFd, V: AsRawFd> Launcher<'a, Started, U, V> {
+            .ioctl(self.kvm, &mut cmd)
+            .map_err(|e| cmd.encapsulate(e))?;
+
+        Ok(())
```
I think this should automatically pin these pages (see #65 for more info). I opened #68 to backport @haraldh's work from sev-iocuddle along with an example of what I'm talking about to implicitly pin the pages.

Feel free to cherry-pick those commits from my branch if you want to make immediate use of them and get them merged as part of this PR. If you end up doing that, be sure to comment on the other PR so we know we can close it.

haraldh commented on src/launch/launcher.rs:
```diff
@@ -71,6 +112,7 @@ impl<'a, U: AsRawFd, V: AsRawFd> Launcher<'a, Started, U, V> {
         LAUNCH_UPDATE_DATA
             .ioctl(self.kvm, &mut cmd)
             .map_err(|e| cmd.encapsulate(e))?;
+        ENC_REG_REGION.ioctl(self.kvm, &KvmEncRegion::new(data))?;
```
The other way round please!
```suggestion
        ENC_REG_REGION.ioctl(self.kvm, &KvmEncRegion::new(data))?;

        LAUNCH_UPDATE_DATA
            .ioctl(self.kvm, &mut cmd)
            .map_err(|e| cmd.encapsulate(e))?;
```

haraldh replied:
Same goes for non-SNP, btw.

haraldh commented on src/launch/launcher.rs:
```diff
@@ -93,6 +135,31 @@ impl<'a, U: AsRawFd, V: AsRawFd> Launcher<'a, Started, U, V> {
+        let mut cmd = Command::from(self.sev, &launch_update_data);
+        SNP_LAUNCH_UPDATE
+            .ioctl(self.kvm, &mut cmd)
+            .map_err(|e| cmd.encapsulate(e))?;
```
```suggestion
        ENC_REG_REGION.ioctl(self.kvm, &KvmEncRegion::new(update.uaddr))?;

        let mut cmd = Command::from(self.sev, &launch_update_data);
        SNP_LAUNCH_UPDATE
            .ioctl(self.kvm, &mut cmd)
            .map_err(|e| cmd.encapsulate(e))?;
```

haraldh reviewed (approved):
I am approving this, although the structs for the ID_BLOCK and ID_AUTH are missing.

Filed an issue https://github.com/enarx/sev/issues/69

haraldh commented on src/launch/mod.rs:
```diff
@@ -192,3 +194,224 @@ impl codicon::Encoder<()> for Measurement {
+        Self {
+            uaddr,
+            policy,
+            ma_en: uaddr.is_some(),
```
This is incorrect... not every guest is a migration agent!

haraldh replied:
Ah, sorry, this is the migration assistent uaddr.

haraldh replied:
That wasn't clear

haraldh replied:
should probably renamed to `ma_uaddr`

tylerfanelli replied:
I see how that can be confusing and agree. Changed it to `ma_uaddr`

haraldh committed:
feat: add `as_mut_vmfd()` to `Launcher`

As we can only do `vm_fd.create_vcpu()` in between
`Launcher::snp_start()` and `Launcher::snp_finish()`, we need access to the
internal virtual machine file descriptor.

haraldh committed:
feat: run a real SNP VM

The code in the VM does a VM termination request via the GHCB MSR.

In the kernel dmesg buffer, the `5.13.0-0.rc1.14.snp.rfc4.1.fc34.x86_64`
kernel gives us the provides exit code:
```
[323835.755893] SEV-ES guest requested termination: 0x1:0x2
```
but returns the vCPU `run()` with `EINVAL`, which is of course
unfortunate.

This also removes dangerous_hw_tests, as it does no harm.

haraldh commented on src/launch/launcher.rs:
```diff
@@ -61,16 +65,57 @@ impl<'a, U: AsRawFd, V: AsRawFd> Launcher<'a, New, U, V> {
+
+    /// Begin the SEV-SNP launch process by creating a Launcher and issuing the
+    /// KVM_SNP_INIT ioctl.
+    pub fn snp_new(kvm: &'a mut U, sev: &'a mut V) -> Result<Self> {
```
So, I made a real running SNP VM here:
https://github.com/haraldh/sev/blob/real_code/tests/snp_launch.rs

For this to work, I needed a `&mut vm_fd` (here named `kvm`).
[This](https://github.com/haraldh/sev/commit/7d2775847a49f3af5c886fc7c30606bc48427a20) patch enabled it.

Patch for the real snp_launch test:
https://github.com/haraldh/sev/commit/eacfd06360b362c9064847cf28b3f0471c3f8f81

haraldh replied:
Btw, I would love to see a rename from `kvm` to `kvm_vm` or `vm_fd` throughout the crate to distinguish it from the `kvm` fd, just to prevent confusion.
